### ISSUE TITLE
Code completion is not "fuzzy"

### ISSUE THREAD

**Issue by mihai-dinu** [ananthakumaran/tide#331]
When submitting issues to this project, please include the following information.

## Checklist

- [x] I have searched both open and closed issues and cannot find a duplicate.
- [x] I can [reproduce the problem](https://github.com/ananthakumaran/tide/wiki/Explanations-about-the-Issue-Reporting-Checklist#use-the-latest-version-of-the-relevant-packages) with the latest version of the relevant packages.
- [x] The problem still occurs after I issued [`M-x tide-restart-server`](https://github.com/ananthakumaran/tide/wiki/Explanations-about-the-Issue-Reporting-Checklist#try-m-x-tide-restart-server) in the buffer where I had the problem.
- [x] I verified that the version and the configuration file path reported by [`M-x tide-verify-setup`](https://github.com/ananthakumaran/tide/wiki/Explanations-about-the-Issue-Reporting-Checklist#check-the-information-reported-by-m-x-tide-verify-setup) are correct.
- [x] If [tide *is reporting*](https://github.com/ananthakumaran/tide/wiki/Explanations-about-the-Issue-Reporting-Checklist#the-behavior-of-cli-tools-agrees-with-your-expectations) an error or warning I think *should not be reported*, I can run `tsc` (and `tslint`, if applicable) without the error or warning I'm seeing in tide.
- [x] If [tide is *not reporting*](https://github.com/ananthakumaran/tide/wiki/Explanations-about-the-Issue-Reporting-Checklist#the-behavior-of-cli-tools-agrees-with-your-expectations) an error or warning I think *should be reported*, `tsc` (or `tslint`, if applicable) reports the error or warning I was expecting to see.
- [x] I am [positive](https://github.com/ananthakumaran/tide/wiki/Explanations-about-the-Issue-Reporting-Checklist#how-to-determine-which-project-the-problem-belongs-to) the problem does not belong to `typescript-mode` or `tsserver`.

## Relevant Version Numbers

- Tide: 20190706.2322
- TypeScript: 3.5.3
- Emacs: 26.2 (latest Doom Emacs from `develop` branch)

## Steps to Reproduce the Bug

In an existing typescript project or in a new one import `fs` and try to check what `dir` related function does the `fs` module have by typing `fs.dir` and waiting on a dropdown list with all the function that contain dir in the title. 

```typescript
import * as fs from 'fs'

fs.dir // waiting on completion dropdown
```

## Expected Behavior

A dropdown list with: readdir, readdirSync, mkdir, mkdirSync, etc

## Actual Behavior

Only one option for `Dirent` in the dropdown, since the tide completion backend seems to only show the functions/classes that **begin** with the sequence of characters you've typed. 
So typing in `fs.read` will show me a dropdown with all the functions that start with `read` like `readfile`, `readdir`, etc

### Notes from issuer

I'm fairly sure this might be a problem with my setup because I couldn't have been the only one that finds this frustrating, if it were a general issue. But I can't for the life of me figure out what the problem is. 

I'm using Doom Emacs with tide-mode and typescript-mode enabled and the `company-tide` as the first entry in the `company-backends` list.

Full list of modes enabled for TS files: 
> Enabled minor modes: +Javascript-Npm +Popup Async-Bytecomp-Package
Auto-Composition Auto-Compression Auto-Encryption Better-Jumper
Better-Jumper-Local Column-Number Company Company-Prescient Diff-Auto-Refine
Display-Line-Numbers Doom-Modeline Dtrt-Indent Eldoc Electric-Indent Evil
Evil-Collection-Unimpaired Evil-Escape Evil-Goggles Evil-Local Evil-Mc
Evil-Snipe Evil-Snipe-Local Evil-Snipe-Override Evil-Snipe-Override-Local
Evil-Surround File-Name-Shadow Flycheck Flycheck-Popup-Tip Font-Lock Format-All
General-Override Global-Company Global-Eldoc Global-Evil-Collection-Unimpaired
Global-Evil-Mc Global-Evil-Surround Global-Flycheck Global-Font-Lock
Global-Git-Commit Global-Magit-File Global-Undo-Tree Highlight-Numbers Hl-Line
Hl-Todo Ivy Ivy-Rich Line-Number Magit-Todos Menu-Bar Mouse-Wheel Npm Persp
Prescient-Persist Projectile Rainbow-Delimiters Recentf Save-Place Savehist
Shell-Dirtrack Show-Paren Size-Indication Smartparens Smartparens-Global Solaire
Solaire-Global Tide Tool-Bar Transient-Mark Undo-Tree Vi-Tilde-Fringe Which-Key
Window-Divider Winner Ws-Butler Ws-Butler-Global Yas
>
> (Information about these minor modes follows the major mode info.)
>
> typescript mode defined in ‘typescript-mode.el’:
> Major mode for editing typescript.

I think this is a problem with `tide` because I've tried using `lsp-mode` with `company-lsp` and I have no problem with the autocompletion there, but lsp is very slow and is not a valid alternative. 

Output from `M-x tide-verify-setup`:
> tsserver version: 3.5.3
>
> config file path: /tmp/tstest/tsconfig.json

`company-backends` variable:
> Value in #\<buffer index.ts\>
(company-tide company-capf company-yasnippet)

**Comment by ananthakumaran** [comment]
tide explicitly filters out results without prefix match `(string-prefix-p prefix (plist-get completion :name) tide-completion-ignore-case)`. I am not sure if it's a bug.

We could consider it as a feature request (disabled by default). We need to define what do we mean by fuzzy match, is it just `contains` or something else. We also need to figure out how this would interact with other features where we might assume completion would match prefix (sorting/auto import on selection)

I think we can only enable this for member completion, especially if we ask tsserver for completions without any prefix, it would return too many result and make everything slow.

**Comment by mihai-dinu** [comment]
>  We need to define what do we mean by fuzzy match, is it just contains or something else.

Yes, I'm referring to just `contains`. So, if the function/class `contains` the `prefix` then show that as a possible option in the dropdown. Pretty much every editor that I've used for typescript has this functionality and I find it useful when you don't really remember the full method name but just some part of it and then you type that in and browse through the results of the completion. 

>I think we can only enable this for member completion, especially if we ask tsserver for completions without any prefix, it would return too many result and make everything slow.

Doesn't `lsp-mode` do this? Ask for completions without a prefix? Because as mentioned before it does show me the all the `fs` functions that contain `dir` (in my example). And from what I know `lsp-mode` uses tsserver as well.

**Comment by ananthakumaran** [comment]
```elisp
(when (and (not (tide-in-string-p)) (not (tide-member-completion-p prefix)))
  (setq file-location (-concat file-location `(:prefix ,prefix))))
```

Currently, tide always asks for a prefix completion unless it's a member completion. Member completion here means something like `object.prefix|`.

This change was made long ago because tsserver used to respond with too many results. Just to check the current status, I disabled prefix and tried completion at the top level. Tsserver replied with 309 KB data. Unfortunately, emacs can't parse it fast enough without freezing the UI see #184.

I think doing fuzzy at the top level is not going to be straightforward. I opened a PR #332 that supports fuzzy completion (member only). Try it out and let me know if it works.

**Comment by mihai-dinu** [comment]
Hi @ananthakumaran 

I'm not sure how to install the package form the PR. What I tried was to download the `tide.el` file from the PR and to `M-x package-install-file` pointing to it. I got the `tide-completion-fuzzy` variable, but setting it to `t` did not have any effect.

**Comment by ananthakumaran** [comment]
```elisp
(add-to-list 'load-path "../repos/tide/")
(require 'tide)
# more config
```
you could clone the repo, checkout the branch and use the above config. Make sure to remove the elpa package (uninstall and delete any compiled files `tide.elc`). I don't know about `package-install-file` specially considering the case where tide is already installed.

**Comment by mihai-dinu** [comment]
I've made some time to test out the PR. 
With `tide-completion-fuzzy 't` I don't see any performance problems in my case. There is a problem though with how fuzzy the completion is. An example:

`fs.dir|` will return `readdir, mkdir, etc` -> correct and what I was expecting
`fs.op|` will return `open, openSync` -> partially correct, I was expecting to see `copyFile, copySync` as well 

And there is also a problem with the fuzzy completion with ignoring case. Example:

```javascript
class test {
   addService() {}
   serviceRemove() {}
}
```

`test.service|` will return `serviceRemove` -> partially correct, was expecting `addService` in the list
`test.Service|` will return `addService, serviceRemove` -> correct and what I was expecting

Basically, the completion should by default be case insensitive when fuzzy and ideally show the closes matches first in the list (in `test.Service|` show the `addService` method before the `serviceRemove` method).

**Comment by ananthakumaran** [comment]
I found an issue with the implementation. company maintains an internal cache and re-uses when we type. For example, if it fetches candidates for `di`, then it will reuse the same candidates for `dir`. The problem is, it filters using `prefix` predicate, not based on `contains`. I couldn't find any way to change the filter other than disabling the [cache](https://github.com/ananthakumaran/tide/pull/332/commits/447ea8878b2522dc641721178e59867c35b95187) altogether. company-lsp seems to implement its own cache as well. If you set `tide-completion-ignore-case` to t, then case will not be considered.

**Comment by paldepind** [comment]
> This change was made long ago because tsserver used to respond with too many results. Just to check the current status, I disabled prefix and tried completion at the top level. Tsserver replied with 309 KB data. Unfortunately, emacs can't parse it fast enough without freezing the UI see #184.

I wonder if the native JSON parsing in 27 changes anything about that? Having some support for fuzzy completion would be really great 😃

**Comment by ananthakumaran** [comment]
btw there is an open PR related to this https://github.com/ananthakumaran/tide/pull/332, but was not merged due to lack of feedback. if you want this feature, suggest trying it out and leave comments there.

**Comment by ananthakumaran** [comment]
#332 is merged. Open a new issue if any of the items are not covered by the PR
